lms/lmsweb/git_service.py

```diff
@@ -0,0 +1,214 @@
+
+        if self._request.path.endswith(upload_pack_command):
+            content_type = 'application/x-git-upload-pack-result'
+            service_command = [
+                upload_pack_command,
+                stateless_rpc,
+                self.project_name,
+            ]
+
+        elif self._request.path.endswith(receive_pack_command):
+            content_type = 'application/x-git-receive-pack-result'
+            service_command = [
+                receive_pack_command,
+                stateless_rpc,
+                self.project_name,
+            ]
+            contain_new_commits = True
+
+        elif self._request.path.endswith('/info/refs'):
+            service_name = self._request.args.get('service')
+            service_command = [
+                service_name,
+                stateless_rpc,
+                advertise_refs,
+                self.project_name,
+            ]
+            content_type = f'application/x-{service_name}-advertisement'
+
+            supported = service_name in allowed_commands
+
+            def format_response_callback(response_bytes: bytes) -> bytes:
+                packet = f'# service={service_name}\n'
+                length = len(packet) + 4
+                prefix = '{:04x}'.format(length & 0xFFFF)
+
+                data = (prefix + packet + '0000').encode()
+                data += response_bytes
+                return data
+
+            format_response = format_response_callback
+
+        else:
+            _logger.error(
+                'Failed to find the git command for route %s',
+                self._request.path,
+            )
+            raise NotImplementedError
+
+        return _GitOperation(
+            response_content_type=content_type,
+            service_command=service_command,
+            supported=supported,
+            format_response=format_response,
+            contain_new_commits=contain_new_commits,
+        )
+
+    def _load_files_from_repository(self) -> typing.List[upload.File]:
+        with tempfile.TemporaryDirectory() as tempdir:
+            proc = subprocess.Popen(  # noqa: S603
+                args=['git', 'clone', self.repository_folder, '.'],
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                cwd=tempdir,
+            )
+            return_code = proc.wait()
+            if return_code != 0:
+                _logger.error(
+                    'Failed to execute command. stdout=%s\nstderr=%s',
+                    proc.stdout.read(), proc.stderr.read(),
+                )
+                raise EnvironmentError
+            to_return = []
+            # remove git internal files
+            shutil.rmtree(os.path.join(tempdir, '.git'))
+            for root, _, files in os.walk(tempdir):
+                for file in files:
+                    upload_file = self._load_file(file, root, tempdir)
+                    to_return.append(upload_file)
```

Comment:
Maybe we should return an Iterator instead?

Reply:
All next interfaces use a list, the linter didn't like it at all.

Reply:
The linter didn't like it 'cause the typing here is typing.List, if you'll replace it with typing.Iterator it'll be OK.

It's not that biggy since you use it only in one place which consumes it all, but usually you'd prefer to serve iterators wherever possible and cast them to list whenever needed